### DISCUSSION TITLE
CScreenDevice private functions

### DISCUSSION THREAD

**Discussion by xlar54** [rsta2/circle#301] [Ideas]
Heya, was wondering if there would be any negative effect to changing some of the private functions to public.  I really would like to use DisplayChar directly so I can set column and row myself.  The escape code method isnt very intuitive and rather cumbersome in my case.  Any concerns about making some of those functions public?

**Comment by rsta2** [comment]
At the moment you should be able to call `CScreenDevice::DisplayChar()` directly, after making it public. But this may change in a future version. The class `CScreenDevice` is meant to be working like a terminal device and directly calling such methods breaks this model.

**Comment by xlar54** [comment]
i guess technically i could just copy and create my own screendevice object for this specific purpose.  yours is a good foundation and a good general purpose class.  btw, appreciate all you do with this.  Im working on a fantasy 65816 computer.  thanks again!

**Reply by rsta2** [reply]
Creating your own custom class based on `CScreenDevice` is a good idea and a common solution, because Circle cannot address all possible application requirements. Thanks for appreciating Circle and good luck for your interesting project!

**Comment by ArnoldUK** [comment]
I take it then that the CScreenDevice should be used for console and character based output and the C2DGraphics device for graphical draws and FrameBuffer writes. I'm not quite sure which device FrameBuffer to use or is best for blitting graphics and bitmaps to the screen ? Is the FrameBuffer shared between both the CScreenDevice and C2DGraphics device ?

**Reply by rsta2** [reply]
`CScreenDevice` emulates a scrolling terminal window with escape control sequences, but provides also a `SetPixel()` method for simple graphics. `C2DGraphics` allows more complex graphics and is probably better for your purpose. I recently added a `DrawText()` method to it on the *develop* branch. Only one of these classes can be used at a time, because both contain a `CBcmFrameBuffer` member object. Some users also used the `CBcmFrameBuffer` class directly and implemented their own graphics routines based on it.

**Comment by ArnoldUK** [comment]
Yes, I have used the `CBcmFrameBuffer` directly and works as expected. Only wanted to confirm with you that the other display classes shared the same `CBcmFrameBuffer`.

I going to look at the develop branch as the `DrawText()` function sounds quite useful along with the `C2DGraphics` device.
I'm mostly using the `CBcmFrameBuffer` directly and sending log output to a `CScreenDevice`.

Cheers @rsta2 once again your replies addressed my questions perfectly.
